planner/core/task.go

```diff
@@ -1035,11 +1033,15 @@ func (t *copTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
 		}
 		ts := tp.(*PhysicalTableScan)
 		prevColumnLen := len(ts.Columns)
-		prevSchema = ts.schema.Clone()
+		prevSchema := ts.schema.Clone()
 		ts.Columns = ExpandVirtualColumn(ts.Columns, ts.schema, ts.Table.Columns)
 		if len(ts.Columns) > prevColumnLen {
 			// Add an projection to make sure not to output extract columns.
-			needExtraProj = true
+			t.needExtraProj = true
+			// Keep originSchema if exists so that we don't overlap it.
+			if t.originSchema == nil {
+				t.originSchema = prevSchema
```

Comment:
Why not remove `prevSchema` variable and replace it here with `ts.schema.Clone()`?

Reply:
ExpandVirtualColumn would change the schema.

Reply:
Ah, thanks for the explanation!